### ISSUE TITLE
slow Roberta pretraining from scratch 

### ISSUE THREAD

**Issue by aabayarea** [pytorch/fairseq#3226]
Hello, 

I am pretraining Roberta from scratch on 64x16GB RAM GPUs on 330 GB (split into 128 partitions) of text but currently, at epoch 32 and pertaining seems to be very slow.  Is this behavior normal?

Thanks in advance for your help. 

TOTAL_UPDATES=2800000000    # Total number of training steps
WARMUP_UPDATES=10000    # Warmup the learning rate over this many updates
PEAK_LR=0.00004          # Peak learning rate, adjust as needed
TOKENS_PER_SAMPLE=512   # Max sequence length
MAX_POSITIONS=512       # Num. positional embeddings (usually same as above)
MAX_SENTENCES=4        # Number of sequences per batch (batch size)
UPDATE_FREQ=32         # Increase the batch size 16x

![image](https://user-images.githubusercontent.com/37912140/107418446-822fad00-6acb-11eb-9bec-488c74f19dd8.png)

**Comment by myleott** [comment]
We'd need more details to help. For example, what was the exact training command? How are the 64 GPUs interconnected to one another? What kind of GPUs are you using? This can be important, since `--fp16` is only effective on Volta-generation GPUs or newer (and may actually be slower on older GPUs).

I see you have `2800000000` updates, but this is wayyy to much. Probably a couple hundred thousand will be sufficient.

Also, `--update-freq 32` is quite high, so it's actually doing 32 forward/backward passes per update. While this does achieve the desired 8192 batch size, in practice we found a batch size of 2048 performed comparably [1]. 

[1] See Table 2 of the OpenReview version of the paper: https://openreview.net/pdf?id=SyxS0T4tvS

**Comment by aabayarea** [comment]
Hello Myle, 

Thanks for your response. I'm not sure what I'm doing wrong here but see below are the hyperparameters.  Any help will be very helpful. 

Hardware = 128 V100 (16GB RAM), 32 machines, and each machine has 4 GPUs (Yes, I doubled the HW)

Data= 340 GB, 2.8 billion docs, processed to have a sliding window of 8 sentences, with a step of 4 sentences. 

With the hyperparameters below, it takes about 1 hour to go through one epoch (shared). I increased the update freq after epoch 42 and I'm currently at epoch 92. Increasing update freq, processes each epoch faster

After epoch 50, the val loss function almost plateaued at 0.413, and perplexity plateaued at 1.33.

I'm training Roberta from scratch. Downstream finetuning such as NER is improving slowly with more pretraining but still worse than SOTA. 

NODE_RANK=$1
TOTAL_UPDATES=2800000000    # Total number of training steps
WARMUP_UPDATES=10000    # Warmup the learning rate over this many updates
PEAK_LR=0.000005          # Peak learning rate, adjust as needed
TOKENS_PER_SAMPLE=512   # Max sequence length
MAX_POSITIONS=512       # Num. positional embeddings (usually same as above)
MAX_SENTENCES=4        # Number of sequences per batch (batch size)
UPDATE_FREQ=128         # Increase the batch size 16x

python -m torch.distributed.launch --nproc_per_node=4 \
    --nnodes=32 --node_rank=$NODE_RANK --master_addr="10.11.129.10" \
    --master_port=65000 \
    $(which fairseq-train) --fp16 $DATA_DIR_0:$DATA_DIR_1:$DATA_DIR_2:$DATA_DIR_3:$DATA_DIR_4:$DATA_DIR_5:$DATA_DIR_6:$DATA_DIR_7:$DATA_DIR_8:$DATA_DIR_9:$DATA_DIR_10:$DATA_DIR_11:$DATA_DIR_12:$DATA_DIR_13:$DATA_DIR_14:$DATA_DIR_15:$DATA_DIR_16:$DATA_DIR_17:$DATA_DIR_18:$DATA_DIR_19:$DATA_DIR_20:$DATA_DIR_21:$DATA_DIR_22:$DATA_DIR_23:$DATA_DIR_24:$DATA_DIR_25:$DATA_DIR_26:$DATA_DIR_27:$DATA_DIR_28:$DATA_DIR_29:$DATA_DIR_30:$DATA_DIR_31:$DATA_DIR_32:$DATA_DIR_33:$DATA_DIR_34:$DATA_DIR_35:$DATA_DIR_36:$DATA_DIR_37:$DATA_DIR_38:$DATA_DIR_39:$DATA_DIR_40:$DATA_DIR_41:$DATA_DIR_42:$DATA_DIR_43:$DATA_DIR_44:$DATA_DIR_45:$DATA_DIR_46:$DATA_DIR_47:$DATA_DIR_48:$DATA_DIR_49:$DATA_DIR_50:$DATA_DIR_51:$DATA_DIR_52:$DATA_DIR_53:$DATA_DIR_54:$DATA_DIR_55:$DATA_DIR_56:$DATA_DIR_57:$DATA_DIR_58:$DATA_DIR_59:$DATA_DIR_60:$DATA_DIR_61:$DATA_DIR_62:$DATA_DIR_63:$DATA_DIR_64:$DATA_DIR_65:$DATA_DIR_66:$DATA_DIR_67:$DATA_DIR_68:$DATA_DIR_69:$DATA_DIR_70:$DATA_DIR_71:$DATA_DIR_72:$DATA_DIR_73:$DATA_DIR_74:$DATA_DIR_75:$DATA_DIR_76:$DATA_DIR_77:$DATA_DIR_78:$DATA_DIR_79:$DATA_DIR_80:$DATA_DIR_81:$DATA_DIR_82:$DATA_DIR_83:$DATA_DIR_84:$DATA_DIR_85:$DATA_DIR_86:$DATA_DIR_87:$DATA_DIR_88:$DATA_DIR_89:$DATA_DIR_90:$DATA_DIR_91:$DATA_DIR_92:$DATA_DIR_93:$DATA_DIR_94:$DATA_DIR_95:$DATA_DIR_96:$DATA_DIR_97:$DATA_DIR_98:$DATA_DIR_99:$DATA_DIR_100:$DATA_DIR_101:$DATA_DIR_102:$DATA_DIR_103:$DATA_DIR_104:$DATA_DIR_105:$DATA_DIR_106:$DATA_DIR_107:$DATA_DIR_108:$DATA_DIR_109:$DATA_DIR_110:$DATA_DIR_111:$DATA_DIR_112:$DATA_DIR_113:$DATA_DIR_114:$DATA_DIR_115:$DATA_DIR_116:$DATA_DIR_117:$DATA_DIR_118:$DATA_DIR_119:$DATA_DIR_120:$DATA_DIR_121:$DATA_DIR_122:$DATA_DIR_123:$DATA_DIR_124:$DATA_DIR_125:$DATA_DIR_126:$DATA_DIR_127  \
    --task masked_lm --criterion masked_lm  \
    --arch roberta_large --sample-break-mode complete --tokens-per-sample $TOKENS_PER_SAMPLE \
    --optimizer adam --adam-betas '(0.9,0.98)' --adam-eps 1e-6 --clip-norm 0.0 \
    --lr-scheduler polynomial_decay --lr $PEAK_LR --warmup-updates $WARMUP_UPDATES --total-num-update $TOTAL_UPDATES \
    --dropout 0.1 --attention-dropout 0.1 --weight-decay 0.01 \
    --batch-size $MAX_SENTENCES --update-freq $UPDATE_FREQ \
    --max-update $TOTAL_UPDATES --log-format simple --log-interval 1 \
    --checkpoint-suffix "_all_notes_window8_" \
    --restore-file   ~/pretrained_models/Roberta/roberta.large.fresh.all.notes/checkpoint_last_all_notes_window8_.pt  \
    --skip-invalid-size-inputs-valid-test \
    --save-dir ~/pretrained_models/Roberta/roberta.large.fresh.all.notes/ \
    --dataset-impl=mmap  --log-format tqdm  --log-interval 1  --tensorboard-logdir ~/logs

![image](https://user-images.githubusercontent.com/37912140/108411013-f9a4c100-71dc-11eb-9f0c-f7d05073bdc8.png)
